core/src/main/java/io/questdb/cutlass/text/ParallelCsvFileImporter.java

```diff
@@ -0,0 +1,1441 @@
+
+        for (int i = 0, n = distinctKeys.size(); i < n; i++) {
+            long key = distinctKeys.getQuick(i);
+            long size = totalSizes.getQuick(i);
+
+            partitionNameSink.clear();
+            dirFormat.format(distinctKeys.get(i), null, null, partitionNameSink);
+            String dirName = partitionNameSink.toString();
+
+            partitions.add(new PartitionInfo(key, dirName, size));
+        }
+    }
+
+    private void processInner() throws TextException {
+        long startMs = getCurrentTimeMs();
+
+        final long fd = TableUtils.openRO(ff, inputFilePath, LOG);
+        try {
+            long length = ff.length(fd);
+            if (length < 1) {
+                throw TextException.$("ignored empty input file [file='").put(inputFilePath).put(']');
+            }
+
+            try (TableWriter writer = parseStructure(fd)) {
+                findChunkBoundaries(length);
+                indexChunks();
+                importPartitions();
+                mergeSymbolTables(writer);
+                updateSymbolKeys(writer);
+                buildColumnIndexes(writer);
+                try {
+                    movePartitions();
+                    attachPartitions(writer);
+                } catch (Throwable t) {
+                    cleanUp(writer);
+                    throw t;
+                }
+            } catch (Throwable t) {
+                cleanUp();
+                throw t;
+            } finally {
+                if (createdWorkDir) {
+                    removeWorkDir();
+                }
+            }
+        } finally {
+            ff.close(fd);
+        }
+
+        long endMs = getCurrentTimeMs();
+        LOG.info().$("import complete [file='").$(inputFileName).$("', time=").$((endMs - startMs) / 1000).$("s").I$();
+    }
+
+    @TestOnly
+    void processTest(Runnable callback) throws SqlException, TextException {
+        if (!lock.tryLock()) {
+            throw SqlException.position(0).put("Another parallel copy command in progress");
+        }
+        try {
+            callback.run();
+            processInner();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void removeWorkDir() {
+        Path workDirPath = tmpPath.of(importRoot).$();
+
+        if (ff.exists(workDirPath)) {
+            if (isOneOfMainDirectories(importRoot)) {
+                throw CairoException.instance(0).put("cannot remove work dir because it points to one of main instance directories [path='").put(workDirPath).put("'] .");
+            }
+
+            LOG.info().$("removing import directory path='").$(workDirPath).$("'").$();
+
+            int errno = ff.rmdir(workDirPath);
+            if (errno != 0) {
+                throw TextException.$("could not remove [path='").put(workDirPath).put("', errno=").put(errno).put(']');
+            }
+        }
+    }
+
+    private boolean isOneOfMainDirectories(CharSequence p) {
```

Comment:
we only check that we're deleting something bad, however we could be overwriting `db` directory or config or anything like this. We should be checking this perhaps on system's startup?

And disallow `..` in csv name ?

Reply:
I added a similar check to PropServerConfiguration that fails if work dir is the same as conf, db root, etc. 
Csv path should be rather innocuous because we only read from it  but use target table name to create and delete work dir - and table name is checked for .. and other characters . 
So far imports allowed setting copy root to "" or null or / and using absolute paths .